File: src/com/castlefrog/agl/agents/ExternalAgent.java
package com.castlefrog.agl.agents;

import java.util.concurrent.CountDownLatch;

import com.castlefrog.agl.Action;
import com.castlefrog.agl.Agent;
import com.castlefrog.agl.Simulator;
import com.castlefrog.agl.State;

public final class ExternalAgent implements Agent {
    /** selected action to return from selectAction */
    private Object action_;
    /** indicates when external program has updated action_ */
    private CountDownLatch actionReady_;

    public ExternalAgent() {
        actionReady_ = new CountDownLatch(1);
    }

    public <S extends State<S>, A extends Action> A selectAction(int agentId, S state, Simulator<S, A> simulator) {
        simulator.setState(state);
        if (simulator.hasLegalActions(agentId)) {
            try {
                actionReady_.await();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            @SuppressWarnings("unchecked")
            A action = (A) action_;
            actionReady_ = new CountDownLatch(1);
            return action;
        }
        return null;
    }

    /**
     * Set the action for this agent.
     */
    public synchronized <A extends Action> void setAction(A action) {
        action_ = action;
        actionReady_.countDown();
    }

    public String getName() {
        return "external";
    }

    @Override
    public String toString() {
        return getName() + " agent";
    }
}

File: src/main/kotlin/com/castlefrog/agl/agents/ExternalAgent.kt
package com.castlefrog.agl.agents

import java.util.concurrent.CountDownLatch

import com.castlefrog.agl.Action
import com.castlefrog.agl.Agent
import com.castlefrog.agl.Simulator
import com.castlefrog.agl.State
import java.util.Optional

class ExternalAgent : Agent {
    /** selected action to return from selectAction  */
    private var action: Any? = null
    /** indicates when external program has updated action_  */
    private var actionReady: CountDownLatch = CountDownLatch(1)

    override fun <S : State<S>, A : Action<A>> selectAction(agentId: Int, state: S, simulator: Simulator<S, A>): Optional<A> {
        simulator.state = state
        actionReady = CountDownLatch(1)
        actionReady.await()
        return Optional.of(this.action as A)
    }

    /**
     * Set the action for this agent.
     */
    @Synchronized fun <A : Action<A>> setAction(action: A) {
        this.action = action
        actionReady.countDown()
    }

    override fun toString(): String {
        return javaClass.simpleName
    }

}
